File: server/inscode/sampler.py
"""Art mode: decide, per module, whether to draw a solid block, a small mark, or
leave the artwork alone.

A scanner samples the *middle* of each module, so a small mark at each centre
carries the data while the artwork stays visible in between. That is also why
brightness is measured at the module centre rather than averaged over the whole
module -- the edges are exactly the part the scanner ignores.

Mirrors web/src/qr/sampler.ts. When a spec arrives with `art.cells` already filled
in, the browser's decisions are used instead of recomputing these -- see svg.py.
"""
import base64
import binascii
import io

from PIL import Image

from .encode import Matrix
from .geometry import finder_cells, finder_centres, structural_cells
from .spec import DARK_OK, LIGHT_OK, MARK_DARK, MARK_LIGHT, SKIP, SOLID_DARK, SOLID_LIGHT, QRSpec
from .svg import canvas_size, logo_rect

#: Samples per module edge. 3 keeps the centre third addressable, as in the original.
SAMPLE_RES = 3
#: Modules of breathing room around each finder disc.
FINDER_GAP = 0.5


def load_image(src: str) -> Image.Image:
    """Decode a data: URI or a filesystem path into RGBA."""
    if src.startswith("data:"):
        try:
            _, b64 = src.split(",", 1)
        except ValueError as e:
            raise ValueError("malformed data URI") from e
        try:
            raw = base64.b64decode(b64)
        except (binascii.Error, ValueError) as e:
            raise ValueError("logo is not valid base64") from e
        return Image.open(io.BytesIO(raw)).convert("RGBA")
    return Image.open(src).convert("RGBA")


def _place(spec: QRSpec, matrix: Matrix, img: Image.Image, background) -> Image.Image:
    """The artwork composited onto a SAMPLE_RES-per-module canvas."""
    total = canvas_size(matrix.size, spec.canvas.quiet_zone)
    px = total * SAMPLE_RES
    base = Image.new("RGBA", (px, px), background)
    rect = logo_rect(spec, matrix.size, (img.width / img.height) if img.height else 1.0)
    if rect:
        x, y, w, h = rect
        scaled = img.resize(
            (max(1, round(w * SAMPLE_RES)), max(1, round(h * SAMPLE_RES))), Image.LANCZOS
        )
        base.alpha_composite(scaled, (round(x * SAMPLE_RES), round(y * SAMPLE_RES)))
    return base


def sample_luminance(spec: QRSpec, matrix: Matrix, img: Image.Image, light: str):
    """Luminance under every module. `light` is what shows through transparent art."""
    base = _place(spec, matrix, img, light)
    flat = Image.new("RGB", base.size, light)
    flat.paste(base, (0, 0), base)
    return flat.convert("L").load(), base.size[0]


def decide_cells(spec: QRSpec, matrix: Matrix, lum=None, width: int = 0) -> bytes:
    """Per-module draw decisions, row-major over the code area."""
    size, qz = matrix.size, spec.canvas.quiet_zone
    skip = finder_cells(size)
    solid = set() if spec.art.loose else structural_cells(matrix.version)

    out = bytearray(size * size)
    for r in range(size):
        for c in range(size):
            i = r * size + c
            if (r, c) in skip:
                out[i] = SKIP  # finders are drawn as rings, separately
                continue
            dark = matrix.get(r, c)
            if (r, c) in solid:
                out[i] = SOLID_DARK if dark else SOLID_LIGHT
                continue
            if lum is None:
                out[i] = MARK_DARK if dark else SKIP
                continue
            L = lum[(c + qz) * SAMPLE_RES + 1, (r + qz) * SAMPLE_RES + 1]
            if dark and L > DARK_OK:
                out[i] = MARK_DARK
            elif not dark and L < LIGHT_OK:
                out[i] = MARK_LIGHT
            else:
                out[i] = SKIP  # the artwork already reads correctly here
    return bytes(out)


def finder_clearance(spec: QRSpec, matrix: Matrix, img: Image.Image) -> list[bool]:
    """Whether each finder disc is free of opaque artwork.

    Tested against the alpha channel rather than the bounding box, so a round logo
    is not punished for its corners.
    """
    base = _place(spec, matrix, img, (0, 0, 0, 0))
    alpha = base.getchannel("A").load()
    px = base.size[0]
    qz = spec.canvas.quiet_zone
    out = []
    for cx, cy in finder_centres(matrix.size):
        dx, dy = (cx + qz) * SAMPLE_RES, (cy + qz) * SAMPLE_RES
        rad = 3.5 * SAMPLE_RES
        clear = True
        for y in range(max(0, int(dy - rad)), min(px, int(dy + rad) + 1)):
            for x in range(max(0, int(dx - rad)), min(px, int(dx + rad) + 1)):
                if alpha[x, y] <= 16:
                    continue
                if (x + 0.5 - dx) ** 2 + (y + 0.5 - dy) ** 2 <= rad * rad:
                    clear = False
                    break
            if not clear:
                break
        out.append(clear)
    return out


def fit_clear_of_finders(spec: QRSpec, matrix: Matrix, img: Image.Image) -> float:
    """Largest artwork scale whose opaque pixels stay clear of the finder discs.

    Binary search on scale, testing alpha against the three finder circles -- so a
    round logo can stay large where a square one has to shrink. Port of
    `_fit_clear_of_finders` from the original script.
    """
    if spec.logo is None:
        return 1.0
    qz = spec.canvas.quiet_zone
    discs = [
        ((cx + qz) * SAMPLE_RES, (cy + qz) * SAMPLE_RES, ((3.5 + FINDER_GAP) * SAMPLE_RES) ** 2)
        for cx, cy in finder_centres(matrix.size)
    ]

    def hits(scale: float) -> bool:
        probe = spec.model_copy(deep=True)
        probe.logo.scale = scale
        base = _place(probe, matrix, img, (0, 0, 0, 0))
        alpha = base.getchannel("A").load()
        px = base.size[0]
        for dx, dy, r2 in discs:
            rad = (3.5 + FINDER_GAP) * SAMPLE_RES
            for y in range(max(0, int(dy - rad)), min(px, int(dy + rad) + 1)):
                for x in range(max(0, int(dx - rad)), min(px, int(dx + rad) + 1)):
                    if alpha[x, y] <= 16:
                        continue
                    if (x + 0.5 - dx) ** 2 + (y + 0.5 - dy) ** 2 <= r2:
                        return True
        return False

    wanted = spec.logo.scale
    if not hits(wanted):
        return wanted
    lo, hi = 0.05, wanted
    for _ in range(14):
        mid = (lo + hi) / 2
        if hits(mid):
            hi = mid
        else:
            lo = mid
    return lo

File: web/src/qr/sampler.ts
/**
 * Art mode: decide, per module, whether to draw a solid block, a small mark, or
 * leave the artwork alone.
 *
 * A scanner samples the *middle* of each module, so a small mark at each centre
 * carries the data while the artwork stays visible in between. That is also why
 * brightness is measured at the module centre and not averaged over the whole
 * module -- the edges are exactly the part the scanner ignores.
 *
 * The result travels in `spec.art.cells`, so the Python renderer draws these
 * decisions rather than reproducing this sampling bit-for-bit. It can still
 * compute its own for direct API use; the two agree to within a threshold.
 */
import { CellKind, DARK_OK, LIGHT_OK, type QRSpec } from './spec'
import type { Matrix } from './encode'
import { cellKey, finderCells, finderCentres, structuralCells } from './geometry'
import { canvasSize, logoRect } from './layout'

/** Samples per module edge. 3 keeps the centre third addressable, as in the original. */
const SAMPLE_RES = 3

export function loadImage(src: string): Promise<HTMLImageElement> {
  return new Promise((resolve, reject) => {
    const img = new Image()
    img.onload = () => resolve(img)
    img.onerror = () => reject(new Error('could not decode the image'))
    img.src = src
  })
}

/**
 * Luminance under every module, at SAMPLE_RES x SAMPLE_RES per module.
 * `light` is what shows through where the artwork is transparent -- on a
 * transparent background that is whatever the code is eventually placed on, which
 * we have to assume is white.
 */
export function sampleLuminance(
  spec: QRSpec,
  matrix: Matrix,
  img: HTMLImageElement,
  light: string,
): Uint8ClampedArray {
  const { quietZone } = spec.canvas
  const total = canvasSize(matrix.size, quietZone)
  const px = total * SAMPLE_RES

  const canvas = document.createElement('canvas')
  canvas.width = px
  canvas.height = px
  const ctx = canvas.getContext('2d', { willReadFrequently: true })!
  ctx.fillStyle = light
  ctx.fillRect(0, 0, px, px)

  const aspect = img.naturalWidth / img.naturalHeight || 1
  const rect = logoRect(spec, matrix.size, aspect)
  if (rect) {
    ctx.imageSmoothingQuality = 'high'
    ctx.drawImage(img, rect.x * SAMPLE_RES, rect.y * SAMPLE_RES, rect.w * SAMPLE_RES, rect.h * SAMPLE_RES)
  }

  const { data } = ctx.getImageData(0, 0, px, px)
  const lum = new Uint8ClampedArray(px * px)
  for (let i = 0, j = 0; i < data.length; i += 4, j++) {
    // Rec. 601 luma, matching PIL's "L" conversion.
    lum[j] = (data[i] * 299 + data[i + 1] * 587 + data[i + 2] * 114) / 1000
  }
  return lum
}

/**
 * Per-module draw decisions, row-major over the *code* area (matrix.size^2).
 */
export function decideCells(
  spec: QRSpec,
  matrix: Matrix,
  lum: Uint8ClampedArray | null,
): Uint8Array {
  const { size, version } = matrix
  const { quietZone } = spec.canvas
  const total = canvasSize(size, quietZone)
  const skip = finderCells(size)
  const solid = spec.art.loose ? new Set<number>() : structuralCells(version)

  const out = new Uint8Array(size * size)
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      const i = r * size + c
      const key = cellKey(r, c)
      if (skip.has(key)) {
        out[i] = CellKind.SKIP // finders are drawn as rings, separately
        continue
      }
      const dark = matrix.get(r, c)
      if (solid.has(key)) {
        out[i] = dark ? CellKind.SOLID_DARK : CellKind.SOLID_LIGHT
        continue
      }
      if (!lum) {
        out[i] = dark ? CellKind.MARK_DARK : CellKind.SKIP
        continue
      }
      // Centre sub-cell of this module, on the quiet-zone-inclusive sample grid.
      const sx = (c + quietZone) * SAMPLE_RES + 1
      const sy = (r + quietZone) * SAMPLE_RES + 1
      const L = lum[sy * total * SAMPLE_RES + sx]
      if (dark && L > DARK_OK) out[i] = CellKind.MARK_DARK
      else if (!dark && L < LIGHT_OK) out[i] = CellKind.MARK_LIGHT
      else out[i] = CellKind.SKIP // the artwork already reads correctly here
    }
  }
  return out
}

/** Run-length encode cells to base64: repeating (kind, count) byte pairs. */
export function encodeCells(cells: Uint8Array): string {
  const bytes: number[] = []
  let i = 0
  while (i < cells.length) {
    const kind = cells[i]
    let run = 1
    while (i + run < cells.length && cells[i + run] === kind && run < 255) run++
    bytes.push(kind, run)
    i += run
  }
  let s = ''
  for (const b of bytes) s += String.fromCharCode(b)
  return btoa(s)
}

export function decodeCells(b64: string, count: number): Uint8Array {
  const bin = atob(b64)
  const out = new Uint8Array(count)
  let at = 0
  for (let i = 0; i + 1 < bin.length; i += 2) {
    const kind = bin.charCodeAt(i)
    const run = bin.charCodeAt(i + 1)
    for (let k = 0; k < run && at < count; k++) out[at++] = kind
  }
  if (at !== count) throw new Error(`art.cells covers ${at} modules, expected ${count}`)
  return out
}

/**
 * Whether each finder disc is free of opaque artwork.
 *
 * Art mode needs this because artwork showing through a finder's middle ring
 * reads as a dark module and kills the lock-on. Tested against the alpha channel,
 * not the bounding box, so a round logo is not punished for its corners.
 */
export function finderClearance(spec: QRSpec, matrix: Matrix, img: HTMLImageElement): boolean[] {
  const { quietZone } = spec.canvas
  const total = canvasSize(matrix.size, quietZone)
  const px = total * SAMPLE_RES
  const canvas = document.createElement('canvas')
  canvas.width = px
  canvas.height = px
  const ctx = canvas.getContext('2d', { willReadFrequently: true })!
  const aspect = img.naturalWidth / img.naturalHeight || 1
  const rect = logoRect(spec, matrix.size, aspect)
  if (!rect) return [true, true, true]
  ctx.drawImage(img, rect.x * SAMPLE_RES, rect.y * SAMPLE_RES, rect.w * SAMPLE_RES, rect.h * SAMPLE_RES)
  const { data } = ctx.getImageData(0, 0, px, px)

  return finderCentres(matrix.size).map(({ cx, cy }) => {
    const dx = (cx + quietZone) * SAMPLE_RES
    const dy = (cy + quietZone) * SAMPLE_RES
    const rad = 3.5 * SAMPLE_RES
    for (let y = Math.max(0, Math.floor(dy - rad)); y < Math.min(px, Math.ceil(dy + rad)); y++) {
      for (let x = Math.max(0, Math.floor(dx - rad)); x < Math.min(px, Math.ceil(dx + rad)); x++) {
        if (data[(y * px + x) * 4 + 3] <= 16) continue
        if ((x + 0.5 - dx) ** 2 + (y + 0.5 - dy) ** 2 <= rad * rad) return false
      }
    }
    return true
  })
}
